### ISSUE TITLE
tensor-autograd/tensor-webgpu: dense full-table Embedding backward + no WebGPU shader/pipeline/buffer caching (4 perf findings)

### ISSUE THREAD

**Issue by johnhenry** [johnhenry/math-plus#100]
Four performance findings across `tensor-autograd` and `tensor-webgpu`. `tensor-autograd`'s finding was verified by direct execution (safe, pure JS). `tensor-webgpu`'s findings are STATIC ANALYSIS ONLY — headless WebGPU init in this environment can hang indefinitely with no GPU present, so no WebGPU code was executed; `packages/tensor-webgpu/test/helpers.ts` itself launches real headless Chrome + Xvfb + CDP to drive `navigator.gpu`, confirming this is the correct way to test it, not something safely done ad hoc.

## 1. `Embedding.forward`'s backward closure materializes the full dense embedding table on every call, regardless of batch size
**`packages/tensor-autograd/src/nn.ts:140-162`**
```ts
const acc: number[][] = Array.from({ length: numEmbeddings }, () => new Array(embeddingDim).fill(0));
```
Allocates and zero-fills the entire embedding table's shape every backward pass, even when only a handful of rows were touched — cost scales with table size, not batch size. Verified: a batch of **3** row-indices into a 50,000×256 embedding table took **~770ms** for forward+backward (dominated by this allocation/accumulate/`Tensor.from(acc.flat())` step), independent of batch size. The module's own doc comment acknowledges the lack of a scatter-add primitive as a known v1 limitation but doesn't call out that the workaround pays for the full table on every step even at batch size 1.

**Fix**: sparse accumulation via a `Map<rowIdx, number[]>` (or scatter directly into a flat typed array only at the touched rows) instead of a dense `numEmbeddings × embeddingDim` allocation per step.

## 2. Shader module + compute pipeline recompiled from scratch on every dispatch (static analysis)
**`packages/tensor-webgpu/src/gpu-runtime.ts:79-80`, `gemm.ts:109-110`, `attention.ts:38-39`** — all three call `device.createShaderModule({code})` + `device.createComputePipeline({layout:"auto",...})` unconditionally on every invocation, zero caching, confirmed via `grep -n "cache\|pool" src/*.ts` returning zero hits. For `gemm.ts`/`attention.ts` the shader source is a fixed constant string (`GEMM_WGSL`, `QKT_WGSL`, `SOFTMAX_WGSL`, `WEIGHTED_SUM_WGSL`) that never changes between calls, so every call recompiles an identical shader/pipeline. `elementwise.ts`'s IR-compiled shaders have the same issue for repeated calls with the same traced expression (the normal case in a training loop). Shader/pipeline compilation is one of the most expensive WebGPU driver operations; paying it per-call in what's meant to be an accelerated hot path defeats much of the purpose. Unlike the GEMM tiling limitation (explicitly documented as deferred in `gemm.ts`'s own doc comment), this gap isn't flagged as a known tradeoff anywhere.

**Fix**: cache compiled shader modules/pipelines keyed by shader source string (or by `(node, numInputs)` for IR-compiled ones).

## 3. GPU buffers allocated and destroyed on every op call, no pooling (static analysis)
**`packages/tensor-webgpu/src/gpu-runtime.ts:23-46`** (`uploadStorageBuffer`/`allocateOutputBuffer`), used by `elementwise.ts:43-50`, `gemm.ts:83-99`, `attention.ts:96-115,159-168,212-231` — every kernel call creates fresh `GPUBuffer`s and destroys them in a `finally` block immediately after that single dispatch. Compounds finding 2: every op call pays both full pipeline-recompilation AND full buffer-allocation/destruction cost.

**Fix**: a size-keyed buffer pool reused across calls of the same shape/dtype.

## 4. Chained GPU ops round-trip through the CPU between every step instead of staying GPU-resident (static analysis)
**`packages/tensor-webgpu/src/attention.ts`** (`runQKT`, `runSoftmax`, `runWeightedSum`) — all three take and return plain `Float32Array` (host memory), not `GPUTensor`/`GPUBuffer`. A standard attention computation chaining them (QKᵀ → softmax → weights·V) does 3 independent upload+dispatch+readback cycles instead of one upload, three dispatches, and one final readback with intermediates staying GPU-resident. The package has a persistent-GPU-buffer abstraction, `GPUTensor` (`device.ts:81-142`), but none of `runGemmWGSL`/`runQKT`/`runSoftmax`/`runWeightedSum`/`runElementwiseWGSL` accept or return it (confirmed via a repo-wide grep — `GPUTensor` referenced nowhere in `src/` except a stale comment in `elementwise.ts:9` pointing at a "GPUTensor-facing wrapper" that doesn't exist in the package). Appears to be planned-but-unbuilt integration work; as-is it multiplies both transfer cost and pipeline-recompilation cost (#2) by the number of chained ops.

**Fix**: have the attention primitives accept/return `GPUTensor` so intermediates stay on-device across the chain.

No correctness bugs were found in either package — gradient correctness, reentrancy/aliasing (double-`backward()`, weight-sharing accumulation), and WebGPU buffer lifecycle (try/finally destroy) all checked out clean.
